### DISCUSSION TITLE
[RRFC] Expose an API for top packages by download count

### DISCUSSION THREAD

**Discussion by dominykas** [npm/rfcs#151] [General]
I'm aware this is not strictly a CLI concern, but @darcyclarke told me to go ahead with this request anyways :)

In the past, there were a couple of spreadsheets shared with the top 1000 most downloaded packages, in some cases privately (much much appreciated!). It would be nice to have these in public, on a regular basis (monthly?)

I used these to scan the node version support that these packages provide as a sort of a measure of overall ecosystem activity and the speed at which it updates to latest node, esp LTS, versions, there's some findings/stats in https://github.com/nodejs/package-maintenance/issues/200. The goal there is to see if we can help the maintainers stay up to date and test in latest node versions.

But I'm sure other people would also find more uses for that list.

I know there already exist a download count API, and it is very flexible, but getting the list of top packages would require querying each existing package one by one and sorting afterwards, which would be quite inefficient.

---

Side note: I'm not sure it is obvious whether issues or discussions should be used - possibly the README needs some updates?

**Comment by ljharb** [comment]
This would be great.

I'd love to see live-ish-updated lists, ideally all with configurable time periods, of:
 - N most downloaded packages
 - N most downloaded package authors
 - N most depended-upon packages
 - N most dev-depended-upon packages

**Comment by 0xdevalias** [comment]
I outlined a bit of what would be required to build something like this given existing tools recently, may be of interest to some:

> > There doesn't seem to be a good API on npm's end to get the names of every package. If there was some sort of daily dump it could be easily done.
> 
> > There's a complete list of all packages at [replicate.npmjs.com/_all_docs](https://replicate.npmjs.com/_all_docs). Maybe I can use that for downloading all package stats in a daily cron job...
> 
> This is basically the most recent/relevant page for updates to that replicate API:
> 
> - https://github.com/orgs/community/discussions/152515
> 
> There seem to be a few issues with it at the moment (recently migrated to a new system), I believe this is the most recent/relevant official update on that + a summary of some of the issues:
> 
> - https://github.com/orgs/community/discussions/152515#discussioncomment-13433941
> 
> But at least in theory, when it's working correctly, you'll be able to download that full snapshot once, and then use the sequence number to just download the updates since that first 'snapshot' you downloaded from:
> 
> - https://replicate.npmjs.com/registry/_changes
> 
> For many usecases, the revision included in those records is useful for matching up with the full package data from one of the other npm API's; but for this use case I don't even think the part matters, as you're mostly just wanting a list of package names; so you could probably:
> 
> - fetch the latest changes since the last snapshot
> - process and de-duplicate the package entries in that snapshot into a 'reduced snapshot' (since you only really care if a package name exists, not how many times/etc)
> - process and de-duplicate that 'reduced snapshot' against the main database (I separated this from the above step since I figured de-duplicating the snapshot against itself is probably cheaper than hitting the DB for each entry; but it may be a negligible cost either way)
> 
> I haven't looked deeply into the download stats API, but I think that would probably actually end up being the 'heavier' part of things; at least if you wanted it to be truly representative (and not just an approximation based on already fetched/cached data). The docs for the download count API are here:
> 
> - https://github.com/npm/registry/blob/main/docs/download-counts.md
> 
> I won't go into that too deeply, since I believe it's what this project uses currently anyway, but probably the most important part of that would be the limitation of bulk queries:
> 
> - https://github.com/npm/registry/blob/main/docs/download-counts.md#bulk-queries
>   - > Important: Scoped packages are not yet supported in bulk queries. So you cannot request `/downloads/point/last-day/@slack/client,@iterables/map` yet.
>   - > Bulk queries are limited to at most 128 packages at a time and at most 365 days of data.
>   - > All other queries are limited to at most 18 months of data. The earliest date for which data will be returned is January 10, 2015.
> 
> Given we can only fetch the details of 128 packages at a time, we can do a bit of 'back of napkin' math to figure out what would be required. Looking at the total number of rows in the replicate API:
> 
> - https://replicate.npmjs.com/registry/_all_docs
>   - `total_rows: 5537482`
> 
> I'm not 100%, but if we assume each of those corresponds with a single npm package, and there aren't duplicates/etc in that count, then we would need `~43,262` queries to fetch all of the download data for those packages for any given time period:
> 
> - `5537482 / 128 = 43,261.578125 ~= 43,262`
> 
> The real number of queries would actually be higher than that, as according to those docs, the bulk query system doesn't work for scoped packages, so each of those would need to be a single query on it's own.
> 
> From a quick search, I didn't find a lot about API rate limits/etc to see how viable it is to make that many requests over a short(ish) period of time; this was the best I found:
> 
> - https://github.com/npm/feedback/discussions/658
>   - https://blog.npmjs.org/post/164799520460/api-rate-limiting-rolling-out.html
> 
> But if it were me implementing it, that number feels like it might be a bit too high for fetching daily stats.. so maybe weekly or monthly would be more viable in that regard.
> 
> That said, there are likely also a lot of packages in that list that aren't really used/downloaded very often; so it would be an interesting exercise to get the full list of package names, and then get their download stats for the last year or so, and then sort/filter based on that. I suspect that there is probably a large number of packages that would have 0 downloads, or at the very least less than a certain low threshold. You could potentially look at the download count on that higher time scale (eg. year, or month maybe) to filter down the full list of packages to a 'reasonably active' list (eg. above a certain download count threshold), and then only fetch the lower time-scale download counts (eg. daily) for that 'reasonably active' list of packages.
> 
> Anyway, that's my /2c of thoughts on this sort of thing based on recent explorations and my own project ponderings/musings.
> 
> _Originally posted by @0xdevalias in https://github.com/pvorb/npm-stat.com/issues/73#issuecomment-2974891511_
